tap-agent/src/tap/receipt_checks_adapter.rs

```diff
@@ -0,0 +1,205 @@
+        _receipt_id: u64,
+    ) -> Result<bool, Self::AdapterError> {
+        unimplemented!();
+    }
+
+    async fn is_valid_allocation_id(
+        &self,
+        allocation_id: Address,
+    ) -> Result<bool, Self::AdapterError> {
+        // TODO: Remove the if block below? Each TAP Monitor is specific to an allocation
+        // ID. So the receipts that are received here should already have been filtered by
+        // allocation ID.
+        if allocation_id != self.allocation_id {
+            return Ok(false);
+        };
+
+        // Check that the allocation ID is not redeemed yet for this consumer
+        match self.tap_allocation_redeemed.value().await {
+            Ok(redeemed) => Ok(!redeemed),
+            Err(e) => Err(AdapterError::AdapterError {
+                error: format!(
+                    "Could not get allocation escrow redemption status from eventual: {:?}",
+                    e
+                ),
+            }),
+        }
+    }
+
+    async fn is_valid_value(&self, value: u128, query_id: u64) -> Result<bool, Self::AdapterError> {
+        let query_appraisals = self.query_appraisals.as_ref().expect(
+            "Query appraisals should be initialized. The opposite should never happen when \
+            receipts value checking is enabled.",
+        );
+        let query_appraisals_read = query_appraisals.read().await;
+        let appraised_value =
+            query_appraisals_read
+                .get(&query_id)
+                .ok_or_else(|| AdapterError::AdapterError {
+                    error: "No appraised value found for query".to_string(),
+                })?;
+        if value != *appraised_value {
+            return Ok(false);
+        }
+        Ok(true)
+    }
+
+    async fn is_valid_gateway_id(&self, gateway_id: Address) -> Result<bool, Self::AdapterError> {
+        let escrow_accounts =
+            self.escrow_accounts
+                .value()
+                .await
+                .map_err(|e| AdapterError::AdapterError {
+                    error: format!("Could not get escrow accounts from eventual: {:?}", e),
+                })?;
+
+        Ok(escrow_accounts
+            .get(&gateway_id)
+            .map_or(false, |balance| *balance > U256::from(0)))
+    }
+}
+
+impl ReceiptChecksAdapter {
+    fn tap_allocation_redeemed_eventual(
+        allocation_id: Address,
+        sender_address: Address,
+        indexer_address: Address,
+        escrow_subgraph: &'static SubgraphClient,
+        escrow_subgraph_polling_interval: u64,
+    ) -> Eventual<bool> {
+        #[derive(serde::Deserialize)]
+        struct AllocationResponse {
+            #[allow(dead_code)]
+            id: String,
+        }
+
+        #[derive(serde::Deserialize)]
+        struct TransactionsResponse {
+            transactions: Vec<AllocationResponse>,
+        }
+
+        timer(Duration::from_secs(escrow_subgraph_polling_interval)).map_with_retry(
```

Comment:
Duration should be in `ms`?

Reply:
Fixed in 8e6460f1b86ba67c1846fde6d5ae3345dbcc047c